pkg/southbound/gnmi/conn.go

```diff
@@ -0,0 +1,187 @@
+	asset := &topoapi.Asset{}
+	err := target.GetAspect(asset)
+	if err != nil {
+		return nil, errors.NewInvalid("target entity %s must have 'onos.topo.Asset' aspect to work with onos-config", target.ID)
+	}
+
+	configurable := &topoapi.Configurable{}
+	err = target.GetAspect(configurable)
+	if err != nil {
+		return nil, errors.NewInvalid("target entity %s must have 'onos.topo.Configurable' aspect to work with onos-config", target.ID)
+	}
+
+	mastership := &topoapi.MastershipState{}
+	err = target.GetAspect(mastership)
+	if err != nil {
+		return nil, errors.NewInvalid("topo entity %s must have 'onos.topo.MastershipState' aspect to work with onos-config", target.ID)
+	}
+
+	tlsOptions := &topoapi.TLSOptions{}
+	err = target.GetAspect(tlsOptions)
+	if err != nil {
+		return nil, errors.NewInvalid("topo entity %s must have 'onos.topo.TLSOptions' aspect to work with onos-config", target.ID)
+	}
+
+	destination := &baseClient.Destination{
+		Addrs:   []string{configurable.Address},
+		Target:  string(target.ID),
+		Timeout: time.Duration(configurable.Timeout),
+	}
+
+	if tlsOptions.Plain {
+		log.Info("Plain (non TLS) connection to ", configurable.Address)
+	} else {
+		tlsConfig := &tls.Config{}
+		if tlsOptions.Insecure {
+			log.Info("Insecure TLS connection to ", configurable.Address)
+			tlsConfig = &tls.Config{InsecureSkipVerify: true}
+		} else {
+			log.Info("Secure TLS connection to ", configurable.Address)
+		}
+		if tlsOptions.CaCert == "" {
+			log.Info("Loading default CA onfca")
+			defaultCertPool, err := certs.GetCertPoolDefault()
+			if err != nil {
+				return nil, err
+			}
+			tlsConfig.RootCAs = defaultCertPool
+		} else {
+			certPool, err := certs.GetCertPool(tlsOptions.CaCert)
+			if err != nil {
+				return nil, err
+			}
+			tlsConfig.RootCAs = certPool
+		}
+		if tlsOptions.Cert == "" && tlsOptions.Key == "" {
+			log.Info("Loading default certificates")
+			clientCerts, err := tls.X509KeyPair([]byte(certs.DefaultClientCrt), []byte(certs.DefaultClientKey))
+			if err != nil {
+				return nil, err
+			}
+			tlsConfig.Certificates = []tls.Certificate{clientCerts}
+		} else if tlsOptions.Cert != "" && tlsOptions.Key != "" {
+			// Load certs given for device
+			tlsConfig.Certificates = []tls.Certificate{setCertificate(tlsOptions.Cert, tlsOptions.Key)}
+		} else {
+			log.Errorf("Can't load Ca=%s , Cert=%s , key=%s for %v, trying with insecure connection",
+				tlsOptions.CaCert, tlsOptions.Cert, tlsOptions.Key, configurable.Address)
+			tlsConfig = &tls.Config{InsecureSkipVerify: true}
+		}
+		destination.TLS = tlsConfig
+	}
+
+	err = destination.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	return destination, nil
+}
+
+func newConn(target *topoapi.Object) *conn {
+	connID := ConnID(uri.NewURI(
+		uri.WithScheme("gnmi"),
+		uri.WithOpaque(string(target.ID))).String())
+	return &conn{
+		id: connID,
+	}
+}
+
+// newGNMIConnection creates a new gNMI connection
+func connect(target *topoapi.Object) (Conn, error) {
```

Comment:
I think this connect function is a little redundant. The problem is I see two ways to create a `Conn` instance: `connect` and `newConn`. The former creates a `grpc.ClientConn` for the `Conn` and the latter doesn’t.